## Decision Trees
### Agenda
1. Introduction to Decision Trees
2. The Decision Tree Algorithms
3. Decision Tree for Classification

### 1. Introduction to Decision Trees
```
* Non-parametric supervised learning method for regression & classification
* It's similar to playing 'dumb charades'.
* A good algorithm will have less & right questions compared to not-so-good one
* The nodes are questions & leafs are prediction
```


In [365]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

### import data

In [366]:
play_data = pd.read_csv('./DSVC/datasets/tennis.csv.txt')

### show data

In [367]:
play_data

outlook  temp humidity  windy play
0      sunny   hot     high  False   no
1      sunny   hot     high   True   no
2   overcast   hot     high  False  yes
3      rainy  mild     high  False  yes
4      rainy  cool   normal  False  yes
5      rainy  cool   normal   True   no
6   overcast  cool   normal   True  yes
7      sunny  mild     high  False   no
8      sunny  cool   normal  False  yes
9      rainy  mild   normal  False  yes
10     sunny  mild   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes
13     rainy  mild     high   True   no

* A decision tree for above data

<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt4.PNG?raw=true" width="600px">

### 2. Decision Tree Algorithm

* Decision Tree is based on (CART) which is advancement of ID3, developed in 1986 by Ross Quinlan.
* ID3 works when feature data & target data both are categorical in nature
* C4.5 is an advancement of ID3, it coverts continues features into categorical features. Then, proceeds with ID3
* CART is based on C4.5, with slight advancement of 'target can be continues'.
* scikit-learn decision trees are based on CART

#### Criterion of creating Decision Tree
* Entropy - Objective of CART is to maximize information gain in each split
* Gini Impurity - If classes are mixed, gini impurity is maximul
##### Both the approaches, yields almost same results. We will discuss algorithm using Entropy


### Entropy of play
* Entropy(play) = – p(Yes) . log2p(Yes) – p(No) . log2p(No)

In [368]:
play_data.play.value_counts()
#查看表格某列中有多少个不同值的快捷方法，并计算每个不同值有在该列中有多少重复值。

yes    9
no     5
Name: play, dtype: int64

### Calculation Entropy

根据上边公式，你需要完成下面代码块中Entropy_Play计算的代码

In [369]:
import math 
numEntries = len(play_data.play)          #返回数据集的行数
labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data.play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
       
Entropy_Play = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries    #选择该标签(Label)的概率
    Entropy_Play -= prob*log(prob, 2)            #利用公式计算 

{'o': 5, 's': 9}


In [370]:
Entropy_Play

0.9402859586706309

你的Entropy_Play正确结果应该为：0.940285

### Information Gain
* The information gain is based on the decrease in entropy after a dataset is split on an attribute. 
* Constructing a decision tree is all about finding attribute that returns the highest information gain (i.e., the most homogeneous branches).
* Gain(S, A) = Entropy(S) – ∑ [ p(S|A) . Entropy(S|A) ]
* We intend to choose the attribute, splitting by which information gain will be the most
* Next step is calculating information gain for all attributes

#### Information Gain on splitting by Outlook
* Gain(Play, Outlook) = Entropy(Play) – ∑ [ p(Play|Outlook) . Entropy(Play|Outlook) ]
* Gain(Play, Outlook) = Entropy(Play) – [ p(Play|Outlook=Sunny) . Entropy(Play|Outlook=Sunny) ] – [ p(Play|Outlook=Overcast) . Entropy(Play|Outlook=Overcast) ]  – [ p(Play|Outlook=Rain) . Entropy(Play|Outlook=Rain) ] 

In [371]:
play_data[play_data.outlook == 'sunny']

outlook  temp humidity  windy play
0    sunny   hot     high  False   no
1    sunny   hot     high   True   no
7    sunny  mild     high  False   no
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

### Calculation Information Gain

你需要完成下面代码块中Entropy_Play_Outlook_Sunny计算的代码

In [372]:
numEntries_s = len(play_data[play_data.outlook == 'sunny'])          #返回数据集的行数

labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data[play_data.outlook == 'sunny'].play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
        
Entropy_Play_Outlook_Sunny = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries_s    #选择该标签(Label)的概率
    Entropy_Play_Outlook_Sunny -= prob*log(prob, 2)            #利用公式计算 

{'o': 3, 's': 2}


In [373]:
Entropy_Play_Outlook_Sunny

0.9709505944546686

你的Entropy_Play_Outlook_Sunny正确结果应该为：0.970951

In [374]:
play_data[play_data.outlook == 'overcast']

outlook  temp humidity  windy play
2   overcast   hot     high  False  yes
6   overcast  cool   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes

In [375]:
# Entropy(Play|Outlook=overcast)
# Since, it's a homogenous data entropy will be 0
numEntries_o = len(play_data[play_data.outlook == 'overcast'])          #返回数据集的行数

labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data[play_data.outlook == 'overcast'].play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
        
Entropy_Play_Outlook_Overcast = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries_o    #选择该标签(Label)的概率
    Entropy_Play_Outlook_Overcast -= prob*log(prob, 2)            #利用公式计算 

{'s': 4}


In [376]:
Entropy_Play_Outlook_Overcast

0.0

In [377]:
play_data[play_data.outlook == 'rainy']

outlook  temp humidity  windy play
3    rainy  mild     high  False  yes
4    rainy  cool   normal  False  yes
5    rainy  cool   normal   True   no
9    rainy  mild   normal  False  yes
13   rainy  mild     high   True   no

In [378]:
numEntries_r = len(play_data[play_data.outlook == 'rainy'])          #返回数据集的行数

labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data[play_data.outlook == 'rainy'].play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
       
Entropy_Play_Outlook_Rain = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries_r    #选择该标签(Label)的概率
    Entropy_Play_Outlook_Rain -= prob*log(prob, 2)            #利用公式计算 

{'s': 3, 'o': 2}


In [379]:
Entropy_Play_Outlook_Rain

0.9709505944546686

你的Entropy_Play_Outlook_Rain正确结果应该为：0.970951

#### Gain on splitting by attribute outlook

根据以下公式计算Gain

In [380]:
#Gain(Play, Outlook) = Entropy(Play) – [ p(Play|Outlook=Sunny) . Entropy(Play|Outlook=Sunny) ] – 
#[ p(Play|Outlook=Overcast) . Entropy(Play|Outlook=Overcast) ] – [ p(Play|Outlook=Rain) . Entropy(Play|Outlook=Rain) ]

Entropy_Play_Outlook = ((Entropy_Play_Outlook_Sunny*(numEntries_s/numEntries))+(Entropy_Play_Outlook_Overcast*(numEntries_o/numEntries))
                        +(Entropy_Play_Outlook_Rain*(numEntries_r/numEntries)))                                                                               
Gain_Outlook = Entropy_Play - Entropy_Play_Outlook

In [381]:
Gain_Outlook

0.2467498197744391

你的Gain(Play,Outlook)正确结果应该为:0.246750

#### Other gains
* Gain(Play, Temperature) - 0.029
* Gain(Play, Humidity) - 0.151
* Gain(Play, Wind) - 0.048

#### Conclusion - Outlook is winner & thus becomes root of the tree
<img src="https://i1.wp.com/sefiks.com/wp-content/uploads/2017/11/tree-v1.png?zoom=1.25&resize=728%2C252&ssl=1" width="600px">

### Time to find the next splitting criteria

In [382]:
play_data[play_data.outlook == 'overcast']

outlook  temp humidity  windy play
2   overcast   hot     high  False  yes
6   overcast  cool   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes

##### Conclusion - If outlook is overcast, play is true

### Let's find the next splitting feature

In [383]:
play_data[play_data.outlook == 'sunny']

outlook  temp humidity  windy play
0    sunny   hot     high  False   no
1    sunny   hot     high   True   no
7    sunny  mild     high  False   no
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

你需要在下面计算Entropy_Play_Outlook_Sunny

In [420]:
# Entropy(Play_Sunny|)
play_data_sunny = play_data[play_data.outlook == 'sunny']
numEntries_s = len(play_data_sunny)          #返回数据集的行数

labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data_sunny.play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
       
Entropy_Play_Outlook_Sunny = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries_s    #选择该标签(Label)的概率
    Entropy_Play_Outlook_Sunny -= prob*log(prob, 2)            #利用公式计算 

In [419]:
Entropy_Play_Outlook_Sunny

0.9709505944546686

你的Entropy_Play_Outlook_Sunny正确结果应该为:0.970951

### Information Gain for humidity

计算湿度信息增益

In [407]:
#Entropy for attribute high = 0, also entropy for attribute normal = 0 
 
Gain_humidity = Entropy_Play_Outlook_Sunny - 0 - 0 

In [409]:
Gain_humidity

0.9709505944546686

你的Information Gain for humidity正确结果应该为:0.970951

### Information Gain for windy
* False -> 3 -> [1+ 2-]
* True -> 2 -> [1+ 1-]

计算Entropy_Wind_False

In [3]:
numEntries_s = len(play_data.sunny)          #返回数据集的行数

labelCounts = {}                          #保存每个标签(Label)出现次数的字典
for featVec in play_data.sunny([play_data_sunny.wind == False]).play:            #对每组特征向量进行统计
    currentLabel = featVec[-1]            #提取标签(Label)信息
    if currentLabel not in labelCounts.keys():            #keys()以列表返回一个字典所有的键。如果标签(Label)没有放入统计次数的字典,添加进去
        labelCounts[currentLabel] = 0
    labelCounts[currentLabel] +=1                     #Label计数
       
Entropy_Wind_False = 0.0                               #经验熵(香农熵)
for key in labelCounts:                          #计算香农熵
    prob = float(labelCounts[key])/numEntries_s    #选择该标签(Label)的概率
    Entropy_Wind_False -= prob*log(prob, 2)            #利用公式计算 

NameError: name 'play_data' is not defined

In [412]:
Entropy_Wind_False

0.9709505944546686

你的Entropy_Wind_False正确结果应该为:0.918296

计算风信息增益

你的Information Gain for windy正确结果应该为:0.019973

### Information Gain for temperature
* hot -> 2 -> [2- 0+]
* mild -> 2 -> [1+ 1-]
* cool -> 1  -> [1+ 0-]

计算热度信息增益

你的Information Gain for windy正确结果应该为:0.570951

#### Conclusion : Humidity is the best choice on sunny branch

<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt3.PNG?raw=true" width="600px">

In [390]:
play_data[(play_data.outlook == 'sunny') & (play_data.humidity == 'high')]

outlook  temp humidity  windy play
0   sunny   hot     high  False   no
1   sunny   hot     high   True   no
7   sunny  mild     high  False   no

In [391]:
play_data[(play_data.outlook == 'sunny') & (play_data.humidity == 'normal')]

outlook  temp humidity  windy play
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

### Splitting the rainy branch

In [392]:
play_data[play_data.outlook == 'rainy']

outlook  temp humidity  windy play
3    rainy  mild     high  False  yes
4    rainy  cool   normal  False  yes
5    rainy  cool   normal   True   no
9    rainy  mild   normal  False  yes
13   rainy  mild     high   True   no

In [393]:
# Entropy(Play_Rainy|)
Entropy_Play_Outlook_Rainy = None

In [394]:
Entropy_Play_Outlook_Rainy

你的Entropy_Play_Outlook_Rainy正确结果应该为:0.970951

### Information Gain for temp
* mild -> 3 [2+ 1-]
* cool -> 2 [1+ 1-]

计算温度信息增益

正确结果为0.020151

### Information Gain for Windy

计算风信息增益

正确结果为0.970951

### Information Gain for Humidity
* High -> 2 -> [1+ 1-]
* Normal -> 3 -> [2+ 1-]

In [395]:
Entropy_Play_Outlook_Rainy_Normal = None

In [396]:
Entropy_Play_Outlook_Rainy_Normal

Entropy_Play_Outlook_Rainy_Normal的正确结果为0.918296

计算湿度信息增益

正确结果为0.019973

### Final Tree

<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt4.PNG?raw=true" width="600px">

### 3. Decision Tree for Classification
* The leaf nodes of decision tree decides the class
* CART will convert features with continues values into categorical values
* Different tree will be generated with same data given in different order

In [397]:
from sklearn.datasets import load_iris
from sklearn.tree import export_graphviz

In [398]:
iris = load_iris()

In [399]:
iris.data[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [400]:
iris.target[:5]

array([0, 0, 0, 0, 0])

请在下方实现交叉熵DecisionTreeClassifier：

In [401]:
# DecisionTreeClassifier entropy

In [402]:
from sklearn.model_selection import train_test_split

In [403]:
trainX,testX,trainY,testY = train_test_split(iris.data, iris.target)

请利用实现的DecisionTreeClassifier进行训练：

In [404]:
def fit(trainX,trainY):
    pass

fit(trainX,trainY)

In [405]:
export_graphviz(dt,'dt.tree')

NameError: name 'dt' is not defined

### Visualizing the tree
* http://www.webgraphviz.com/

* Criteria - Entropy
<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt6.PNG?raw=true">

在下方写代码来对测试集进行预测：

In [ ]:
def predict(testX):
    pass
predict(testX)

#### Feature Importances
* Important features will be higher up the tree
* We can use this techniques to identify important features

#### Visualizing Decision Decision Boundry

In [ ]:
from sklearn.datasets import make_blobs

In [ ]:
X,Y = make_blobs(n_features=2, n_samples=1000, cluster_std=.8, centers=4, random_state=6)

In [ ]:
plt.scatter(X[:,0],X[:,1],c=Y,s=5, cmap='viridis')

利用基尼系数的决策树实现：

In [ ]:
# DecisionTreeClassifier gini

进行训练

In [ ]:
def fit(trainX,trainY):
    pass

fit(X,Y)

In [ ]:
plot_step = 0.2
x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1
y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, plot_step),
                         np.arange(y_min, y_max, plot_step))

### Calculation Outcome

你需要完成下面代码块中预测值的计算：

In [ ]:
def predict():
    pass
outcome = None

In [ ]:
xx.shape

In [ ]:
plt.scatter(X[:,0],X[:,1],c=Y,s=5,cmap='viridis')
plt.scatter(xx.ravel(),yy.ravel(),c=outcome,s=1,alpha=1, cmap='viridis')
plt.show()

![](img/questions-01.png)